cognite/extractorutils/unstable/core/checkin_worker.py

```diff
@@ -0,0 +1,359 @@
+"""Check-in worker for reporting errors and task updates to the CDF Integrations API."""
+
+import sys
+from collections.abc import Callable
+from logging import Logger
+from secrets import SystemRandom
+from threading import RLock
+from time import sleep
+
+from requests import Response
+
+from cognite.client import CogniteClient
+from cognite.client.exceptions import CogniteAPIError, CogniteAuthError, CogniteConnectionError
+from cognite.extractorutils.threading import CancellationToken
+from cognite.extractorutils.unstable.configuration.models import ConfigRevision
+from cognite.extractorutils.unstable.core._dto import (
+    CheckinRequest,
+    CheckinResponse,
+    JSONType,
+    MessageType,
+    StartupRequest,
+    TaskUpdate,
+)
+from cognite.extractorutils.unstable.core._dto import (
+    Error as DtoError,
+)
+from cognite.extractorutils.unstable.core.errors import Error
+from cognite.extractorutils.util import now
+
+DEFAULT_SLEEP_INTERVAL = STARTUP_BACKOFF_SECONDS = 30.0
+MAX_ERRORS_PER_CHECKIN = MAX_TASK_UPDATES_PER_CHECKIN = 1000
+rng = SystemRandom()
+
+
+class CheckinWorker:
+    """
+    A worker to manage how we report check-ins to the Integrations API.
+
+    This will help us:
+    1. Ensure that we don't report any check-ins before the start up is reported.
+    1. Manage how we batch errors and task updates.
+    2. Manage how we handle retries and backoff.
+    """
+
+    _start_lock = RLock()
```

Comment:
The name `start_lock` is a little misleading, since it appears to be a lock around the `errors` and `task_updates` collections more than anything else.